xdsl/backend/csl/lowering/convert_stencil_to_csl.py

```diff
@@ -0,0 +1,304 @@
+    IntegerAttr,
+    IntegerType,
+    MemRefType,
+    ModuleOp,
+    Signedness,
+    StringAttr,
+    TensorType,
+)
+from xdsl.dialects.csl import (
+    ComptimeStructType,
+    ConcatStructOp,
+    ConstStructOp,
+    CslModuleOp,
+    GetColorOp,
+    ImportModuleConstOp,
+    LayoutOp,
+    MemberCallOp,
+    ModuleKind,
+    ModuleKindAttr,
+    ParamOp,
+    SetRectangleOp,
+    SetTileCodeOp,
+    SignednessCastOp,
+)
+from xdsl.dialects.scf import For, Yield
+from xdsl.dialects.stencil import ApplyOp
+from xdsl.ir import Attribute, Block, MLContext, Operation, Region
+from xdsl.passes import ModulePass
+from xdsl.utils.hints import isa
+
+
+class TranslationContext:
+    grid_dim: tuple[int, ...] = (0, 0, 0)
+    pattern: int
+    program_module: CslModuleOp
+    layout_module: CslModuleOp
+    program_sym_name: StringAttr
+
+    program_params: dict[str, ParamOp] = {}
+    layout_params: dict[str, ParamOp] = {}
+    program_imports: dict[str, ImportModuleConstOp] = {}
+    layout_imports: dict[str, ImportModuleConstOp] = {}
+    program_vars: dict[str, Operation] = {}
+    layout_vars: dict[str, Operation] = {}
+
+    def add_param_to_program(self, param: ParamOp):
+        self.program_module.body.block.add_op(param)
+        self.program_params[param.param_name.data] = param
+
+    def add_param_to_layout(self, param: ParamOp):
+        self.layout_module.body.block.add_op(param)
+        self.layout_params[param.param_name.data] = param
+
+    def add_import_module_to_program(self, lib: ImportModuleConstOp):
+        self.program_module.body.block.add_op(lib)
+        self.program_imports[lib.module.data] = lib
+
+    def add_import_module_to_layout(self, lib: ImportModuleConstOp):
+        self.layout_module.body.block.add_op(lib)
+        self.layout_imports[lib.module.data] = lib
+
+    def add_params_to_program(self, *params: ParamOp):
+        for param in params:
+            self.add_param_to_program(param)
+
+    def add_params_to_layout(self, *params: ParamOp):
+        for param in params:
+            self.add_param_to_layout(param)
+
+    def add_import_modules_to_program(self, *lib: ImportModuleConstOp):
+        for l in lib:
+            self.add_import_module_to_program(l)
+
+    def add_import_modules_to_layout(self, *lib: ImportModuleConstOp):
+        for l in lib:
+            self.add_import_module_to_layout(l)
+
+    def add_var_to_program(self, name: str, op: Operation):
+        self.program_module.body.block.add_op(op)
+        self.program_vars[name] = op
+
+    def add_var_to_layout(self, name: str, op: Operation):
+        self.layout_module.body.block.add_op(op)
+        self.layout_vars[name] = op
+
+
+def setup_program_module(ctx: TranslationContext) -> None:
+    ctx.add_params_to_program(
+        ParamOp.get("stencil_comms_params", ComptimeStructType()),
+        ParamOp.get("memcpy_params", ComptimeStructType()),
+        ParamOp.get("z_dim", IntegerType(16, Signedness.SIGNED)),
```

Comment:
I'd recommend not using signedness unless you *really* need an unsigned integer!

Reply:
dito for all other integer variables

Reply:
I'd caution against that - most of what is happening here is setting up module (library) imports, calling module functions, and using that to set params for the `pe.csl`. It might just be harder to reverse-engineer it in the printer.

Reply:
From the point of view of the printer using signed or signless is exactly the same (they both get printed as signed).

Reply:
we might need to look into that, some functions have specific requirements and may only accept one of the two